tools/fm-editor/FeatureModelEditor.cpp

```diff
@@ -0,0 +1,312 @@
+void FeatureModelEditor::createTreeContextMenu(const QPoint &Pos) {
+  auto Index = TreeView->indexAt(Pos);
+  if (Index.isValid()) {
+    FeatureTreeItem *Item =
+        static_cast<FeatureTreeItem *>(Index.internalPointer())
+            ->child(Index.row());
+    Item->contextMenu(TreeView->mapToGlobal(Pos));
+  }
+}
+
+/// Load the selected file into the textedit and mark the sources of the
+/// selected feature
+///
+/// \param RelativePath  path to the source file relative to the repository path
+void FeatureModelEditor::loadSource(const QString &RelativePath) {
+  Ui->textEdit->clear();
+  auto SourcePath = Repository + "/" + RelativePath;
+  QFile File(SourcePath);
+  if (File.exists()) {
+    File.open(QFile::ReadOnly | QFile::Text);
+    QTextStream ReadFile(&File);
+    Ui->textEdit->setText(ReadFile.readAll());
+    std::vector<vara::feature::FeatureSourceRange> Locations{};
+    std::copy_if(
+        CurrentFeature->getLocationsBegin(), CurrentFeature->getLocationsEnd(),
+        std::back_inserter(Locations), [&RelativePath](auto const &Loc) {
+          return RelativePath.toStdString() == Loc.getPath();
+        });
+    for (auto &Location : Locations) {
+      markLocation(Location);
+    }
+  }
+}
+void setCursorLineAndColumn(QTextCursor &Cursor, int Line, int Col,
+                            QTextCursor::MoveMode Mode) {
+  QTextBlock const B = Cursor.document()->findBlockByLineNumber(Line);
+  Cursor.setPosition(B.position() + Col, Mode);
+}
+
+/// Mark the given SourceRange with the given Format
+///
+/// \param Fmt Format to mark with
+/// \param Cursor  Cursor
+/// \param Location Location to mark
+void FeatureModelEditor::markLocation(
+    vara::feature::FeatureSourceRange &Location) const {
+  QTextCharFormat Fmt;
+  Fmt.setBackground(Qt::darkYellow);
+  QTextCursor Cursor(Ui->textEdit->document());
+  Cursor.movePosition(QTextCursor::Start, QTextCursor::MoveAnchor);
+  setCursorLineAndColumn(Cursor, int(Location.getStart()->getLineNumber()) - 1,
+                         int(Location.getStart()->getColumnOffset()) - 1,
+                         QTextCursor::MoveAnchor);
+  setCursorLineAndColumn(Cursor, int(Location.getEnd()->getLineNumber()) - 1,
+                         int(Location.getEnd()->getColumnOffset()) - 1,
+                         QTextCursor::KeepAnchor);
+  Cursor.setCharFormat(Fmt);
+}
+
+/// Load a sourcefile to then add a location from it
+void FeatureModelEditor::addSourceFile() {
+  if (!Repository.isEmpty()) {
+    QString const Path =
+        QFileDialog::getOpenFileName(this, tr("Select Source File"), Repository,
+                                     tr("C Files (*.c *c++ *.h)"));
+    Ui->sources->addItem(Path.mid(Repository.length()));
+    Ui->sources->setCurrentIndex(Ui->sources->count() - 1);
+  }
+}
+
+/// Add the user selected Part of the textedit as a source for the active
+/// Feature
+void FeatureModelEditor::addSource() {
+  auto *TextEdit = Ui->textEdit;
```

Comment:
```suggestion
  if(Ui->sources->currentText().isEmpty()){
    return;
  }

  auto *TextEdit = Ui->textEdit;
```

Avoid adding "empty" location